### DISCUSSION TITLE
How do you achieve "Resolution of cyclic references of any complexity."

### DISCUSSION THREAD

**Discussion by jzabroski** [activej/activej#249] [Q&A]
I found your library today while researching RPC SerDe libraries. Of all your claims, this was the least typical.

Typically, coming from an object/relational mapping framework background, the way cyclic references are resolved is through Floyd-Warshall and "uniquing" the object graph.  But I get the sense you are _also_ doing something clever in the implementation, and was wondering if you could point me to where that code is, so I can read it.  I'm also curious then if you have tests that test the complexity of the object graph, and how you generate those large object graphs.  I typically generate them through de-composing the problem into smaller problems of creating sub-graphs, and stitching graphs together until I get really huge graphs.

**Comment by eduard-vasinskyi** [comment]
Hi, @jzabroski

I’m not sure that I understood you correctly. I assume you are talking about cyclic reference resolution in `SerializerBuilder`. We achieve this by storing an identity hash map that keeps track of all the serialized objects and corresponding indexes. In case the object is serialized more than once, we simply serialize an index of that object. The code is located [here](https://github.com/activej/activej/blob/v5.4.3/core-serializer/src/main/java/io/activej/serializer/impl/SerializerDefReference.java).Tests for serialization of references can be found [here](https://github.com/activej/activej/blob/v5.4.3/core-serializer/src/test/java/io/activej/serializer/SerializerReferenceTest.java).

As of v5.5 we have moved reference serialization functionality to the `StreamCodec` facility, the code for which can be found [here](https://github.com/activej/activej/blob/86d506f491940ed15c93091eae45af8f31c9c223/core-serializer/src/main/java/io/activej/serializer/stream/StreamCodecs.java#L756-L787).
